compiler/data/kphp-json-tags.cpp

```diff
@@ -403,7 +403,7 @@ FieldJsonSettings merge_and_inherit_json_tags(const ClassMemberInstanceField &fi
 
   // for 'public int $id;' — no default and non-nullable type — set 'required' so that decode() would fire unless exists
   // it could be overridden with `@kphp-json required = false`
-  if (field.type_hint && !field.var->init_val && !does_type_hint_allow_null(field.type_hint)) {
+  if (field.type_hint && !field.var->init_val && !field.var->had_user_assigned_val && !does_type_hint_allow_null(field.type_hint)) {
```

Comment:
And here just
```suggestion
  if (field.type_hint && field.var->uninited_flag && !does_type_hint_allow_null(field.type_hint)) {
```
?

Reply:
Technically speaking inited/not inited value dichotomy is not the problem. The problem is that we can't distinguish default type values (0 for int, 0.0 for float and so on) and same user-provided value.

Reply:
I checked your proposal, it is not enough.
`uninited_flag` allows both `public string $password = '';` and `public string $password;` which is incorrect. Without ` = '';` there must be error.

Reply:
I would change `had_user_assigned_val` into `init_val_was_optimised`, because if it wasn't optimised than this flag would be false.
But up to you